### PR TITLE
Story/CIT-3

### PR DESCRIPTION
**Pull request by namrathaov**
Populate fields on edit page dynamically based on citation type.

### CODE REVIEW

**Comment by namrathaov** [comment]
Retrieved fields related to item type and displayed accordingly.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/EditItemController.java [hunk 1]
```diff
@@ -43,8 +47,31 @@ public String showPage(Authentication authentication, Model model, CitationForm
...
+     */
+    @RequestMapping("/auth/group/{zoteroGroupId}/items/{itemId}/editItem")
+    public ResponseEntity<List<String>> updateItemType(Authentication authentication, Model model, CitationForm form, @PathVariable("zoteroGroupId") String zoteroGroupId, @PathVariable("itemId") String itemId,
+            @RequestParam(value = "itemType") ItemType itemType) {
+        ICitation citation = citationManager.getCitation((IUser)authentication.getPrincipal(), zoteroGroupId, itemId);
+        model.addAttribute("zoteroGroupId", zoteroGroupId);
+        model.addAttribute("citation", citation);
+        if (!model.containsAttribute("form")) {
+            model.addAttribute("form", form);
+        }
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/EditItemController.java]: What are lines 64-69 for?

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/EditItemController.java [hunk 2]
```diff
@@ -43,8 +47,31 @@ public String showPage(Authentication authentication, Model model, CitationForm
...
+        List<String> changedFields = new ArrayList<>();
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/EditItemController.java]: bad variable name; those fields have not been changed, they are just the fields that belong to the specified itemType.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/EditItemController.java [hunk 3]
```diff
@@ -43,8 +47,31 @@ public String showPage(Authentication authentication, Model model, CitationForm
...
+            .forEach(f -> fields.add(f.getFilename()));
+        model.addAttribute("fields", fields);
         return "auth/group/items/item/edit";
     }
+
+    
+    /**
+     * Method to retrieve all fields filtered by item type.
+     */
+    @RequestMapping("/auth/group/{zoteroGroupId}/items/{itemId}/editItem")
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/EditItemController.java]: bad mapping name; this method does not edit an item.

**Comment by namrathaov** [comment]
Refactored code.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/EditItemController.java [hunk 4]
```diff
@@ -43,8 +47,24 @@ public String showPage(Authentication authentication, Model model, CitationForm
...
+        model.addAttribute("fields", fields);
         return "auth/group/items/item/edit";
     }
+
+    
+    /**
+     * Method to retrieve all fields filtered by item type.
+     */
+    @RequestMapping("/auth/items/fields")
+    public ResponseEntity<List<String>> getFieldsByItemType(Authentication authentication, @RequestParam(value = "itemType") ItemType itemType) {
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/EditItemController.java]: I would make the itemType request parameter a path variable (to make it a little bit more rest-like)

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/editItem.jsp [hunk 5]
```diff
@@ -133,7 +135,7 @@ let removeAuthor = function removeAuthor(e) {
 <td>Citation Type</td>
 <td>
 <c:set var="enumValues" value="<%=edu.asu.diging.citesphere.core.model.bib.ItemType.values()%>"/>
-<form:select path="itemType" data-show-icon="true" class="form-control selectpicker">
+<form:select id="itemType" path="itemType" data-show-icon="true" class="form-control selectpicker">
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/editItem.jsp]: why is the id of the form "itemType"?

**Comment by jdamerow** [comment]
ok to test

**Comment by jdamerow** [comment]
Jenkins, test this please.

**Comment by jdamerow** [comment]
Let's also have a little spinner while the fields are changed (it doesn't take very long but to indicate that something is happening), and give the label column a width so that the columns don't jump around when a different item type is changed.

**Comment by namrathaov** [comment]
[CIT-3] Fixed variable name and added load icon.

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/editItem.jsp [hunk 6]
```diff
@@ -185,81 +194,115 @@ ${author.lastName}<c:if test="${not empty author.firstName}">, ${author.firstNam
...
 <tr>
-<td>Call Number</td>
+<td width="20%">Call Number</td>
 <td><form:input path="callNumber" type="text" class="form-control" placeholder="Call Number" value="${not empty form.callNumber ? form.callNumber : citation.callNumber}" /></td>
 </tr>
+</c:if>
+<c:if test="${fn:contains(fields, 'rights') }">
 <tr>
-<td>Rights</td>
+<td width="20%">Rights</td>
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/editItem.jsp]: should be enough to give only one td 20% (one that is always shown), or if we want to change the width, we have to change every single td.

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/editItem.jsp [hunk 7]
```diff
@@ -300,3 +343,32 @@ ${author.lastName}<c:if test="${not empty author.firstName}">, ${author.firstNam
...
+		$('#displayMessage').replaceWith("<div id='displayMessage' class='alert-info row'>"+
+			"<span class='glyphicon glyphicon-refresh spinning'></span>"+
+				"Loading fields... </div>");
+		$('#displayMessage').show();
+		$.ajax({
+			url : '<c:url value="/auth/items/'+itemType+'/fields" />',
+			type : 'GET',
+			success: function(changedFields){
+				$('form input').parent().closest('tr').hide();
+				var i = 0;
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/editItem.jsp]: why are you defining 'i' outside of the loop?

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/editItem.jsp [hunk 8]
```diff
@@ -300,3 +343,32 @@ ${author.lastName}<c:if test="${not empty author.firstName}">, ${author.firstNam
...
+			},
+			error: function(){
+				$('#displayMessage').replaceWith("<div id='errorDiv' class='alert-danger row'>" +
+					"<i class='glyphicon glyphicon-remove-sign'></i>" +
+						"Error loading the form fields</div>");
+				$('#displayMessage').delay(2000).fadeOut();
+			}
+		});
+	})
+});
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/editItem.jsp]: let's make the display message to fit with the rest of the layout, e.g. either grey out the screen or have a blocking modal popup or something.

**Comment by namrathaov** [comment]
Changed the message to be displayed in a modal.

**Pull request merged**